### PR TITLE
[4.2] Fix modal fields

### PR DESCRIPTION
**Pull request by okonomiyaki3000**
Pull Request for Issue #11805 .

### Summary of Changes

New (abstract) field class `ModalField` to serve as base for most modal field classes.
New web component `joomla-field-modal` 
Updated `modal_category`, `modal_contact`, `modal_article`, `modal_menu`, `modal_newsfeed` to use `ModalField` as base class.

### Testing Instructions

These field types are most easily testable on menu item pages. For example, the 'Single Article' menu item type has a `modal_article` field. 

Since the problem with these field types before was that they did not work properly in subforms, test them in subforms. The `fields/checkboxes` plugin has a subform in it, you can easily alter that to include fields of these types. 

### Expected result

Modal behavior should be basically the same as in J3. You can open modals to select, edit, or create items. You can clear the field of all values. 

When using modal fields in a repeatable subform, fields on newly created rows should work just as well as those that were on the page when initially loaded. 

### Actual result

Before the patch, modal fields in repeatable subforms had trouble (#11805 is about `modal_article` but really the problem applies to all of them). It's also possible that all modal fields were broken in J4 anyway.

After the patch, all of these things should work.

### Documentation Changes Required

Maybe not.

### CODE REVIEW

File: media/system/webcomponents/js/joomla-field-modal.es6.js [hunk 1]
```diff
@@ -0,0 +1,578 @@
...
+
+			if (this.elements.buttonSelect)
+			{
+				this.elements.buttonSelect.classList[id ? 'add' : 'remove']('sr-only');
+			}
+			if (this.elements.buttonNew)
+			{
+				this.elements.buttonNew.classList[id ? 'add' : 'remove']('sr-only');
+			}
+			if (this.elements.buttonEdit)
```
[the remark this answers is not in the document]
Comment (okonomiyaki3000) [reply media/system/webcomponents/js/joomla-field-modal.es6.js]: If we're going to be checking for stuff like this, better put a `.editorconfig` file in the root.

File: media/system/webcomponents/js/joomla-field-modal.es6.js [hunk 2]
```diff
@@ -0,0 +1,578 @@
...
+			}
+
+			if (this.elements.fieldId.getAttribute('data-required') == '1')
+			{
+				document.formvalidator.validate(this.elements.fieldId);
+				document.formvalidator.validate(this.elements.fieldTitle);
+			}
+		}
+	});
+})(customElements, Joomla, jQuery);
```
[the remark this answers is not in the document]
Comment (okonomiyaki3000) [reply media/system/webcomponents/js/joomla-field-modal.es6.js]: Bootstrap's modal is implemented as a jQuery plugin. You telling me I need to do a vanilla version of it? I mean... I will, but...

**Comment by dgrammatiko** [comment]
Great work @okonomiyaki3000 !!!
One note here: this file : `media/system/webcomponents/js/joomla-field-modal.es6.js` which is the source of the custom element needs to be moved here: https://github.com/joomla/joomla-cms/tree/4.0-dev/build/webcomponents/js and then you need to run the command `node build.js --compilecejs`

The reason for this is that the `class` on custom elements is not properly transpiled from babel. At some point we need to merge the es6 and the custom elements building scripts...

File: media/system/webcomponents/js/joomla-field-modal.es6.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (okonomiyaki3000) [reply media/system/webcomponents/js/joomla-field-modal.es6.js]: Also, why are switching to spaces now? Is it only for js files? That hardly makes any sense if php is still tabs. Also I vote for tabs.

**Comment by okonomiyaki3000** [comment]
@dgrammatiko Ah, thanks! I had a feeling something was wrong.

**Comment by okonomiyaki3000** [comment]
I guess I also need to add it to package.json in the appropriate places?

File: build/webcomponents/js/field-modal/field-modal.js [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+    processResult(id = '', title = '', catid = '', object = '', url = '', language = '') {
+      // Default values.
+      id       = id || '';
+      title    = title || '';
+      catid    = catid || '';
+      object   = object || '';
+      url      = url || '';
+      language = language || '';
+
+      this.elements.fieldId.value    = id || '';
```
[the remark this answers is not in the document]
Comment (okonomiyaki3000) [reply build/webcomponents/js/field-modal/field-modal.js]: Some of these rules are just dumb.

File: build/webcomponents/js/field-modal/field-modal.js [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+     * @param   {String}  [id='']        Item id
+     * @param   {String}  [title='']     Item title
+     * @param   {String}  [catid='']     Category id
+     * @param   {String}  [object='']    Object
+     * @param   {String}  [url='']       Url
+     * @param   {String}  [language='']  Language
+     *
+     * @return  {void}
+     */
+    processResult(id = '', title = '', catid = '', object = '', url = '', language = '') {
```
[the remark this answers is not in the document]
Comment (okonomiyaki3000) [reply build/webcomponents/js/field-modal/field-modal.js:544]: しょうがないね

File: build/webcomponents/js/field-modal/field-modal.js [hunk 5]
```diff
@@ -0,0 +1,541 @@
...
+          iframe.removeEventListener('load', frameLoaded);
+
+          let iframeDoc = iframe.contentDocument,
+            idField = iframeDoc.getElementById(idFieldId),
+            titleField = iframeDoc.getElementById(titleFieldId);
+
+          if (idField && idField.value != '0') {
+            selected = [idField.value, titleField && titleField.value];
+
+            // If Save & Close (save task), submit the edit close action (so we don't have checked out items).
```
[the remark this answers is not in the document]
Comment (okonomiyaki3000) [reply build/webcomponents/js/field-modal/field-modal.js]: yea..but it's comment so...

File: build/webcomponents/js/field-modal/field-modal.js [hunk 6]
```diff
@@ -0,0 +1,554 @@
...
+          ' ',
+          this.getAttribute('text-button-edit'),
+        );
+
+        this.elements.buttonEdit.classList.add('btn-secondary');
+        this.elements.buttonEdit.classList[hasValue ? 'remove' : 'add']('sr-only');
+        this.elements.buttonEdit.addEventListener('click', () => this.modalEdit(), true);
+      }
+
+      if (this.allow & ALLOW_CLEAR) {
```
[the remark this answers is not in the document]
Comment (okonomiyaki3000) [reply build/webcomponents/js/field-modal/field-modal.js:220]: Bwahahahaha! Get a life, hound! This is absolutely expected, ffs...

File: build/webcomponents/js/field-modal/field-modal.js [hunk 7]
```diff
@@ -0,0 +1,565 @@
+((customElements, Joomla, $) => {
```
[the remark this answers is not in the document]
Comment (okonomiyaki3000) [reply build/webcomponents/js/field-modal/field-modal.js:2]: Not even one? These rules are stupid.

File: build/webcomponents/js/field-modal/field-modal.js [hunk 8]
```diff
@@ -0,0 +1,567 @@
...
+        Joomla.renderModalField.call(this);
+      } else {
+        this.render();
+      }
+    }
+
+    /* eslint no-bitwise: ["error", { "allow": ["|"] }] */
+    /* jshint bitwise: false */
+    get allow() {
+      return (this.getAttribute('allow-new') === 'true' ? ALLOW_NEW : 0) |
```
[the remark this answers is not in the document]
Comment (okonomiyaki3000) [reply build/webcomponents/js/field-modal/field-modal.js:89]: How can I turn off this utterly stupid rule?

**Comment by okonomiyaki3000** [comment]
This dog is provoking me. 

I'm a big fan of linters actually but I strongly disagree with some of the choices in how eslint has been set up. Also, the generated files should not be linted at all.

**Comment by wilsonge** [comment]
This one escaped me - can we sync this back up please?

**Comment by infograf768** [comment]
@okonomiyaki3000 
Can you solve conflicts and be nice to hound ? 😄

**Comment by okonomiyaki3000** [comment]
@infograf768 I'm going to be a bit too busy to get back to it for probably at least a month maybe more. Also I'm not enthusiastic about new js coding standards. Makes me not really want to write js for joomla anymore.

**Comment by gnanakeethan** [comment]
@dgrammatiko  I am a newbie in the J4.0 space. Lets see how things go. I worked on 1.5-3.5 versions. Never been into J! codebase for a long time.  I hope I can work things out. 

@infograf768 thanks for linking the issues. :smile: :+1:

**Comment by uglyeoin** [comment]
Is this ready to test yet?

**Comment by okonomiyaki3000** [comment]
@uglyeoin Nope. Well, if I remember correctly, it should work as it's supposed to. The only problem is that Hound complains about...something. Coding style violations maybe? It's been a while so I don't exactly remember but it's too tedious because I wasn't able to test these violations locally, I just had to upload here and wait for the result then go fix it and repeat. Also, the coding standard Joomla has decided to go with for JS seems pretty awful. I don't remember all the points about it that bothered me but, for example, it forbids the use of bitwise operators. WTF kind of rule is that? Get out of here with that nonsense.

**Comment by coolcat-creations** [comment]
Is it possible to test it now? :-)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20788">issues.joomla.org/tracker/joomla-cms/20788</a>.</sub>

**Comment by okonomiyaki3000** [comment]
@coolcat-creations I have not touched it in a long time and I will not until I can run these hound tests locally which I have also not tried to do for a long time. If someone else wants to pick it up, I would not be sad.

**Comment by Ruud68** [comment]
Hi before creating a new issues, just asking if this is related.
When in Menu > single article > create new article > modal comes up with new article form.
The published and featured are malformed / not working.

Same for editing a plugin in a modal, all btn btn-group radio buttons do not work any more (they switch value, but the new value is only displayed after refresh page.
screen shot 2020-06-19 at 08 07 50<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20788">issues.joomla.org/tracker/joomla-cms/20788</a>.</sub>

**Comment by roland-d** [comment]
@okonomiyaki3000 Do you have any interest in fixing this or should the PR be closed?

**Comment by Ruud68** [comment]
A issue that is not fixed should not be closed.

**Comment by Quy** [comment]
@roland-d See https://github.com/joomla/joomla-cms/pull/20788#issuecomment-612613210

@Ruud68 If it is no longer being maintained, then close the PR and reopen the issue.

**Comment by flo-the-cat** [comment]
I have tested this item :red_circle: unsuccessfully on 8aa4424c86adb3bf26b0e667611ffe92ea4002d1<br /><br />The file marked for modification does not exist: administrator/components/com_categories/Field/Modal/CategoryField.php<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20788">issues.joomla.org/tracker/joomla-cms/20788</a>.</sub>

**Comment by dgrammatiko** [comment]
@okonomiyaki3000 do you mind if I clone this PR to rebase it?

**Comment by Quy** [comment]
@dgrammatiko See his last reply https://github.com/joomla/joomla-cms/pull/20788#issuecomment-612613210

**Comment by laoneo** [comment]
@dgrammatiko do you mind to make a new pr?

**Comment by dgrammatiko** [comment]
Last time we were discussing about this one (basically with @Fedik https://github.com/joomla/joomla-cms/discussions/32473) we conclude that there some known bottlenecks and I don't think I have a solution for all aspects of this PR.
For example (quoting @Fedik here but I also agree) 

```
The interaction between the iframe content and the parent document need to be event based,
without hardcoded callbacks, that often missed somewhere. Basically enough just 2 event:
` joomla:close-popup`, `joomla:select-content` (with data to select).
```

This is platform wide change with some hard B/C breaks, so I don't think this could happen in a minor version, maybe 5.0 (?) but if there's a decision to move towards a better solution in the next minor, I could spend some time to do a POC

**Comment by dgrammatiko** [comment]
BTW dialog is supported everywhere, so there's no need for bootstrap for the modals anymore. Just saying...

**Comment by laoneo** [comment]
I'm closing this one as it is abandoned and has too many hound issues. Till somebody takes it over or starts from scratch, I'v opened an issue #37327.

**Comment by richard67** [comment]
See also discussion https://github.com/joomla/joomla-cms/discussions/32473

**Pull request closed without merging**